ionoscloud/data_source_distribution.go

```diff
@@ -0,0 +1,189 @@
+													Type: schema.TypeSet,
+													Elem: &schema.Schema{
+														Type: schema.TypeString,
+													},
+													Computed: true,
+												},
+												"allow_list": {
+													Type: schema.TypeSet,
+													Elem: &schema.Schema{
+														Type: schema.TypeString,
+													},
+													Computed: true,
+												},
+											},
+										},
+									},
+									"rate_limit_class": {
+										Type:        schema.TypeString,
+										Description: "Rate limit class that will be applied to limit the number of incoming requests per IP.",
+										Computed:    true,
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		Timeouts: &resourceDefaultTimeouts,
+	}
+}
+
+func dataSourceDistributionRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(services.SdkBundle).CdnClient
+
+	idValue, idOk := d.GetOk("id")
+	domainValue, domainOk := d.GetOk("domain")
+
+	id := idValue.(string)
+	domain := domainValue.(string)
+
+	if idOk && domainOk {
+		diags := diag.FromErr(errors.New("id and domain cannot be both specified in the same time"))
+		return diags
+	}
+	if !idOk && !domainOk {
+		diags := diag.FromErr(errors.New("please provide the distribution id or domain"))
+		return diags
+	}
+
+	var distribution cdn.Distribution
+	var err error
+
+	if idOk {
+		/* search by ID */
+		distribution, _, err = client.DistributionsApi.DistributionsFindById(ctx, id).Execute()
+		if err != nil {
+			diags := diag.FromErr(fmt.Errorf("an error occurred while fetching the distribution with ID %s: %w", id, err))
+			return diags
+		}
+	} else {
+		var results []cdn.Distribution
+
+		distributions, _, err := client.DistributionsApi.DistributionsGet(ctx).Execute()
+		if err != nil {
+			diags := diag.FromErr(fmt.Errorf("an error occurred while fetching container distributions: %w", err))
+			return diags
+		}
+
+		results = *distributions.Items
+		if domainOk {
+			partialMatch := d.Get("partial_match").(bool)
+
+			log.Printf("[INFO] Using data source for container registry by domain with partial_match %t and domain: %s", partialMatch, domain)
+
+			if distributions.Items != nil && len(*distributions.Items) > 0 {
+				var distributionsByDomain []cdn.Distribution
+				for _, distributionItem := range *distributions.Items {
+					if distributionItem.Properties != nil && distributionItem.Properties.Domain != nil &&
+						(partialMatch && strings.Contains(*distributionItem.Properties.Domain, domain) ||
+							!partialMatch && strings.EqualFold(*distributionItem.Properties.Domain, domain)) {
+						distributionsByDomain = append(distributionsByDomain, distributionItem)
+					}
+				}
+				if distributionsByDomain != nil && len(distributionsByDomain) > 0 {
+					results = distributionsByDomain
+				} else {
+					return diag.FromErr(fmt.Errorf("no distribution found with the specified criteria: domain = %v", domain))
+				}
+			}
```

Comment:
(just a nitpick) these nested conditions are a bit hard to read, as well as being deeply nested and in a large function

```
if distributions.Items != nil && len(*distributions.Items) > 0 {
	var distributionsByDomain []cdn.Distribution
	for _, item := range *distributions.Items {
		if isValidDomain(item.Properties, domain, partialMatch) {
			distributionsByDomain = append(distributionsByDomain, item)
		}
	}
	if len(distributionsByDomain) > 0 {
		results = distributionsByDomain
	} else {
		return diag.FromErr(fmt.Errorf("no distribution found with the specified criteria: domain = %v", domain))
	}
}

func isValidDomain(properties *cdn.Properties, domain string, partialMatch bool) bool {
	if properties == nil || properties.Domain == nil {
		return false
	}
	if partialMatch {
		return strings.Contains(*properties.Domain, domain)
	}
	return strings.EqualFold(*properties.Domain, domain)
}
```